### DISCUSSION TITLE
Cross Compile for ARMv5 (Feroceon 88FR131)

### DISCUSSION THREAD

**Discussion by amsound** [Barracuda09/SATPI#206] [General]
Hi, I have an old MOI+ box with a TBS6985 DVB-S card in. Ideally would like this to run SATPI however I'm not very confident building with Linux, particularly cross compiling. TBS have provided a tool which I've downloaded from their website and put in /opt/cross on a Debian VM.

Now I'm a bit stuck as to how I can apply this to a SATPI build. Would anyone mind providing some pointers? Should I be able to build for this device with the Broadcom toolchains?

Linux moi 2.6.35.14 #6 Tue May 12 03:05:38 SGT 2015 armv5tel GNU/Linux
Get this with cat /proc/cpuinfo
Processor : Feroceon 88FR131 rev 1 (v5l)
BogoMIPS : 1599.07
Features : swp half thumb fastmult edsp
CPU implementer : 0x56
CPU architecture: 5TE
CPU variant : 0x2
CPU part : 0x131
CPU revision : 1

**Comment by Barracuda09** [comment]
Which version of gcc is this?

**Reply by amsound** [reply]
Looking into the cross compiler, I think 4.3.2

**Reply by Barracuda09** [reply]
With that version it will most certainly not work.
It is a pretty old system and kernel version.

I do not know if it will work (or be easy to make it work) with the Broadcom toolchain (or other ARM toolchain).
The lib dependencies will make it not easy. or it needs to be build statically.

**Reply by amsound** [reply]
Sad face. Any other options you could suggest?

**Comment by Barracuda09** [comment]
Try this if you like: [satpi_ARMV5_1.6.2.177~g410b47f_static.zip](https://github.com/Barracuda09/SATPI/files/13875293/satpi_ARMV5_1.6.2.177.g410b47f_static.zip)

First run `./satpi --version` to see if this works

**Reply by amsound** [reply]
No dice - "Illegal instruction"

**Reply by Barracuda09** [reply]
That is probably because the Broadcom toolchain uses eabi-hf (hard-float) and this uses eabi (soft-float)

**Reply by amsound** [reply]
Thank you for trying. If I was using their tool (https://sourceforge.net/projects/moiplus/files/Files/cross_compile_tool/) would you be able to provide instruction on how to adjust the build process so I could give it a go? Thanks for your help so far.

**Reply by Barracuda09** [reply]
That toolchain will not work, it has gcc 4.3.x that will not do C++17. You need at least gcc 8.3.x
So try to find an Debian package with a toolchain that has that. And I can maybe help you further with building SatPI. It involves changes envARM.sh and CPU_FLAGS for you toolchain and CPU type (ARMV5te)

Maybe in time I will try to look at it myself.

**Reply by amsound** [reply]
Thanks Marc, if you did get a chance that would be amazing. I will do some googling...

**Comment by Saentist** [comment]
@Barracuda09 can yo try with this
https://github.com/sc0ty/nsa310-kernel

My Zyxel Nas have same CPU

```shell
== Welcome at FFP version 0.7/arm, running on NSA310_ZyXEL NSA310, firmware V4.70(AFK.3) ==
root@NSA310:~# lscpu
Architecture:          armv5tel
Byte Order:            Little Endian
CPU(s):                1
On-line CPU(s) list:   0
root@NSA310:~# cat /proc/cpuinfo
Processor       : Feroceon 88FR131 rev 1 (v5l)
BogoMIPS        : 1199.30
Features        : swp half thumb fastmult edsp
CPU implementer : 0x56
CPU architecture: 5TE
CPU variant     : 0x2
CPU part        : 0x131
CPU revision    : 1

Hardware        : Feroceon-KW
Revision        : 0000
Serial          : 0000000000000000
root@NSA310:~# cat /proc/version
Linux version 2.6.31.8 (root@BuildMachine) (gcc version 4.3.2 (sdk3.3-ct-ng-1.4.1) ) #2 Fri Mar 11 17:35:20 CST 2016
```

Guess if TVHeadend can be build and SATPI can.
Linux Kernel 6.6.3 Kirkwood package and Debian rootfs
https://forum.doozan.com/read.php?2,12096

**Reply by amsound** [reply]
This is what I was hoping re Tvheadend which is what currently runs on this device, albeit an old version and lots of other rubbish which it can't really cope with any more!

**Reply by amsound** [reply]
Without much knowledge I amended the 'envARM.sh' document and tried to run make with the TBS Cross Complier and got this:
```
/opt/cross/bin/arm-mv5sft-linux-gnueabi-c++ -c  -I src -std=c++17 -Werror=vla -Wall -Wextra -Winit-self -Wshadow -pthread  -O2 -s -DNDEBUG src/Version.cpp -o obj/Version.o
make: /opt/cross/bin/arm-mv5sft-linux-gnueabi-c++: No such file or directory
make: *** [Makefile:233: obj/Version.o] Error 127
```
And the file is there!

**Reply by Barracuda09** [reply]
what does this show:
`/opt/cross/bin/arm-mv5sft-linux-gnueabi-c++ --version`

**Reply by amsound** [reply]
```
-bash: /opt/cross/bin/arm-mv5sft-linux-gnueabi-c++: No such file or directory
```
or if I go to the directory and run --version I get 
```
arm-mv5sft-linux-gnueabi-c++: command not found
```

**Reply by Barracuda09** [reply]
what about:
`ls /opt/cross/bin`

**Reply by amsound** [reply]
```
arm-mv5sft-linux-gnueabi-addr2line	arm-mv5sft-linux-gnueabi-gdbtui
arm-mv5sft-linux-gnueabi-addr2name.awk	arm-mv5sft-linux-gnueabi-gfortran
arm-mv5sft-linux-gnueabi-ar		arm-mv5sft-linux-gnueabi-gprof
arm-mv5sft-linux-gnueabi-as		arm-mv5sft-linux-gnueabi-jcf-dump
arm-mv5sft-linux-gnueabi-c++		arm-mv5sft-linux-gnueabi-ld
arm-mv5sft-linux-gnueabi-c++filt	arm-mv5sft-linux-gnueabi-nm
arm-mv5sft-linux-gnueabi-cc		arm-mv5sft-linux-gnueabi-objcopy
arm-mv5sft-linux-gnueabi-cpp		arm-mv5sft-linux-gnueabi-objdump
arm-mv5sft-linux-gnueabi-ct-ng.config	arm-mv5sft-linux-gnueabi-populate
arm-mv5sft-linux-gnueabi-g++		arm-mv5sft-linux-gnueabi-ranlib
arm-mv5sft-linux-gnueabi-gcc		arm-mv5sft-linux-gnueabi-readelf
arm-mv5sft-linux-gnueabi-gcc-4.3.2	arm-mv5sft-linux-gnueabi-run
arm-mv5sft-linux-gnueabi-gccbug		arm-mv5sft-linux-gnueabi-size
arm-mv5sft-linux-gnueabi-gcj		arm-mv5sft-linux-gnueabi-strings
arm-mv5sft-linux-gnueabi-gcov		arm-mv5sft-linux-gnueabi-strip
arm-mv5sft-linux-gnueabi-gdb
```

**Reply by Barracuda09** [reply]
This:
`/opt/cross/bin/arm-mv5sft-linux-gnueabi-g++ --version`

**Reply by amsound** [reply]
```
-bash: /opt/cross/bin/arm-mv5sft-linux-gnueabi-g++: No such file or directory
```

**Reply by amsound** [reply]
Weirdly this shows something but only when in the directory:
```
/opt/cross/arm-mv5sft-linux-gnueabi/bin# c++ --version
c++ (Debian 10.2.1-6) 10.2.1 20210110
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Reply by Barracuda09** [reply]
That is most likely not your cross-compiler, what does this show on that same place:
`which c++`

**Reply by amsound** [reply]
```
/usr/bin/c++
```

**Reply by amsound** [reply]
Sorry, I am trying but it's not something I'm practised at. Thank you for your help so far.

**Comment by Barracuda09** [comment]
But what doe this show:
`ls /opt/cross/arm-mv5sft-linux-gnueabi/bin`

Here your path is `/opt/cross/arm-mv5sft-linux-gnueabi/bin`
> Weirdly this shows something but only when in the directory:
> 
> ```
> /opt/cross/arm-mv5sft-linux-gnueabi/bin# c++ --version
> c++ (Debian 10.2.1-6) 10.2.1 20210110
> Copyright (C) 2020 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> ```

**Reply by amsound** [reply]
```
ar  as	c++  g++  gcc  gfortran  ld  nm  objcopy  objdump  ranlib  strip
```

Are you able to try this and see if it works for you?
https://sourceforge.net/projects/moiplus/files/Files/cross_compile_tool/moiplus_corss_compile_tool.zip/download

**Reply by Barracuda09** [reply]
First, one more thing to try:
`/opt/cross/arm-mv5sft-linux-gnueabi/bin/c++ --version`

**Reply by amsound** [reply]
```
-bash: /opt/cross/arm-mv5sft-linux-gnueabi/bin/c++: No such file or directory
```

**Reply by Barracuda09** [reply]
Mmm... probably you "installed" the cross compiler as a different user (root/sudo) with different rights as you are running --version.

Try to 'install' it in your home directory.

**Reply by amsound** [reply]
How do I 'install'? Move it there? I'm running as root.

**Reply by Barracuda09** [reply]
what about this:
`sudo /opt/cross/arm-mv5sft-linux-gnueabi/bin/c++ --version`

**Reply by amsound** [reply]
```
sudo: unable to execute /opt/cross/arm-mv5sft-linux-gnueabi/bin/c++: No such file or directory
```
Should I be using /opt/cross/bin or /opt/cross/arm-mv5sft-linux-gnueabi/bin?

**Reply by Barracuda09** [reply]
I have downloaded it, this is useless. It is from 2009. This does not work.

**Reply by amsound** [reply]
Ha wonderful. Sorry about this.

Are you able to try @Saentist 's idea at all?

**Reply by Barracuda09** [reply]
That is about a new kernel version.

**Reply by amsound** [reply]
So absolutely no way of getting SATPI on this device?

**Reply by Barracuda09** [reply]
Try this [satpi_ARMV5_1.6.2.177~g410b47f_static.zip](https://github.com/Barracuda09/SATPI/files/13919546/satpi_ARMV5_1.6.2.177.g410b47f_static.zip) and run:
`./satpi --version`

**Reply by amsound** [reply]
```
[root@moi /root]# ./satpi --version
FATAL: kernel too old
Aborted
```

**Reply by Barracuda09** [reply]
add this were satpi bin is [ld-linux.so.3.zip](https://github.com/Barracuda09/SATPI/files/13920051/ld-linux.so.3.zip) and run:
./ld-linux.so.3 ./satpi --version

**Reply by amsound** [reply]
```
[root@moi /root]# ./ld-linux.so.3 ./satpi --version
FATAL: kernel too old
Aborted
```

**Comment by amsound** [comment]
Hi Marc, I found another cross compiler which looks like a previous user has got to work - could I ask how best to adapt this to build satpi as simply as possible?

https://www.dropbox.com/scl/fi/6wum31cgayi0jydagcp81/x-tools-moi.tar.bz2?rlkey=780nwdnw7c8r943m8g6ornebb&dl=0

Thanks!

**Reply by amsound** [reply]
Not sure if the "arm-unknown-linux-gnueabi-c++filt" is causing issues due to the 'filt' postfix?

**Reply by Barracuda09** [reply]
This tool-chain uses `arm-unknown-linux-gnueabi-cpp` but this is version 4.9.x this will most likely not work. Unfortunately SatPI uses features of C++ that are probably not supported by this compiler.

**Reply by amsound** [reply]
Oh, never mind! Thanks for trying
